File: src/java/magma/app/rule/Rule.java
package magma.app.rule;

import magma.api.result.Result;
import magma.app.Node;
import magma.app.error.CompileError;

public interface Rule {
    default Result<Node, CompileError> parse(String s) {
        throw new UnsupportedOperationException();
    }

    default Result<String, CompileError> generate(Node node) {
        throw new UnsupportedOperationException();
    }
}

File: src/c/magma/app/rule/Rule.c
import magma.api.result.Result;import magma.app.Node;import magma.app.error.CompileError;struct Rule{
	Result<Node, CompileError> parse(any* _ref_, String input);
	Result<String, CompileError> generate(any* _ref_, Node node);
	struct Impl{
		struct Impl new(){
			struct Impl this;
			return this;
		}
	}
}